### ISSUE TITLE
Alternative for when-directive

### ISSUE THREAD

**Issue by tkroesch** [Polymer/lit-html#695]
I just noticed that the "when" direcitve was removed from version 0.13.0 to 0.14.0.
Why was it removed and what can I use instead?

Thanks for your help! ;-)

**Comment by stramel** [comment]
The `when` directive essentially just did some caching for you. 

Previously, you would use it like so

```js
import { when } from 'lit-html/directives/when';
 let checked = false;
html`
  ${when(checked, () => html`Checkmark is checked`, 
         () => html`Checkmark is not checked`)}
`;
```

and now you use the `cache` directive like so

```js
import { cache } from 'lit-html/directives/cache';
 let checked = false;
html`
  ${cache(checked ? html`Checkmark is checked` : html`Checkmark is not checked`)}
`;
```

**Comment by ruphin** [comment]
`when` combined two different concepts: _conditional rendering_ and _template caching_. The approach in 0.14.0 is that we just expose the primitives, and you can combine these things on your own.

-----

#### Conditional rendering

Conditional rendering in 0.14.0 works by using plain JavaScript features. The most common method is by using the [ternary operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Conditional_Operator):

```js
const template = (condition) => html`
  <div>
    ${
      condition ?
        html`<h2>It's true!<h2>` :
        html`<p>Not really</p>`
    }
  </div>
`;
```

You can use any other JavaScript language feature like `if` conditionals or calling another function or whatever to determine what you render:

```js
const pickSomething = choice => {
  if (choice === 'bold' ) {
    return html`<b>Fat text</b>`
  } else if (choice === 'italic' ) {
    return html`<i>Super slanty</i>`
  } else {
    return html`<span>So plain</span>`
  }
}

const subFunctionTemplate = (choice) => html`
  <div>
    ${ subFunction(choice) }
  <div>
`;
```


You can use an empty template ```html`` ``` or empty string `''` to render nothing. In the next release you can also use the `nothing` export to render nothing (with pretty much the same effect).

```js
const template = (visible) => html`
  <div>
    ${ visible ? html`<h1>I'm here<h1>` : '' }
  </div>
`;
```

-----

#### Template Caching

In all the examples above, when you switch between rendering different templates, the previous content is destroyed, and the new template is rendered from scratch. Sometimes you want to be able to switch between templates but have the old content "cached" so you can switch back and forth without causing a complete re-render of the template. This is what the new `cache` directive is for.

```js
const template = (visible) => html`
  <div>
    ${ cache( visible ? html`<p>I'm cached<p>` : html`` )}
  </div>
`;
```

Just wrap any of the above statements in `cache()` and it will cache any templates that are rendered in that position. Note that although it may seem that caching is always better than not caching there is a performance overhead. Using `cache` will make your application consume more memory, and your application will generally be more efficient without `cache`. I advise you to not use it unless the size of the template is very large _and_ you expect to switch the rendered content often.

------

You can see a working version of all the examples above in this pen: https://codepen.io/ruphin/pen/XoXNBR?editors=1010

Maybe some of this material is useful somewhere in the docs @arthurevans

**Comment by ruphin** [comment]
JavaScript ternary expressions or other conditional statements are conditionally executed, so you do not get any reference errors when you do this:

```js
const template = (loading) => html`
  <div>
    ${
      loading ? 
        html`<p>Loading...<p>` : 
        html`does not use ${ data } until loading is false` )
    }
  </div>
`;
```

Non-executed branches of conditionals are not evaluated by JavaScript at all:

```js
# This gives no errors or anything
true ? 'Something' : UndefinedVariable.nonExistantProperty
```

`when` needed the additional functional wrapping `() => html..` because both branches of the conditional were passed as arguments to a function. JavaScripts evaluates all expressions that are function arguments by default, regardless of how they are used within the function.

**Comment by CaptainCodeman** [comment]
Except

```
let data = null
html`
${cache(loading ? html`loading...` : html`Welcome to ${data.title}`)}
`;
```

... now errors and stops the template rendering. That was precisely the reason that when used functions to return the templates, so they weren't evaluated.

**Comment by ruphin** [comment]
Your example does not error as long as the invariant `loading ^ (data !== undefined)` is valid. In other words, as long as `data` is defined when `loading` is false, you cannot get an error.

**Comment by ruphin** [comment]
If you use an example where you evaluate a branch that leads to a reference error, it will throw a reference error. The example you used switches between each template, so it will evaluate the falsy branch of the conditional, and lead to a null reference error. This would happen just the same if you were using `when`.  Behold: https://codepen.io/ruphin/pen/KbVyKg?editors=1010

If you execute a branch of code that references off of `null` or `undefined`, your code will break. It's up to the programmer to ensure that this doesn't happen. `when` never protected you from that.

What when _did_, is ensure that only one branch is evaluated, and not both. It did so by requiring that you pass both branches as a function, so it could execute only one depending on the provided condition. With JavaScript ternaries or conditionals, you don't have to do that, because they _do this by default_. In other words, the `() =>` wrappings that you had with `when` was a workaround to a shortcoming of `when`, not an added feature.

TLDR; The following two statements are functionally identical, regardless of the content of the templates:

```js
when(condition, () => html`...`, () => html`...`)

cache(condition ? html`...` :  html`...`)
```

**Comment by CaptainCodeman** [comment]
You are correct, I just picked the one where it was toggling the value, duh!

Anyways, now my unsafeHTML is broken ...

**Comment by stramel** [comment]
@tkroesch Feel free to join the #lit-html channel on the Polymer slack https://polymer.slack.com/
